Dailies performance fix - fixes #8756

joe-salomon2 opened the pull request:
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes https://github.com/HabitRPG/habitica/issues/8756

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

For recurring Daily tasks, the [moment-recur](https://github.com/c-trimm/moment-recur) plugin was slowing down calculations for the next due date. This is a known issue with the plugin:

> Important Note: These functions may be very inefficient/slow. They work by attempting to match every date from the start of a range until the desired number of dates have been generated. So if you attempt to get 10 dates for a rule that matches once a year, it will run the match function for ~3650 days.

The logic was rewritten to use [momentjs'](https://momentjs.com/docs/#/manipulating/add/) adding functions to calculate the upcoming due dates.


[//]: # (Put User ID in here - found in Settings -> API)

----
UUID: fc28e9dc-9c84-4ee3-be6e-b4dffa54f558

missingno5678 committed:
Changed recurring logic to not use moment-recur plugin for performance reasons

TheHollidayInn commented:
Thanks for the PR! There isn't an issue with matching and checking for recurrence - just for calculation the next due. So, we don't want to check the "matching" logic. Please only change the nextDue calculation.

Also, can you post your time measurements to show performance improvements?

Great work!

TheHollidayInn commented on website/common/script/cron.js:
```diff
@@ -110,81 +110,95 @@ export function shouldDo (day, dailyTask, options = {}) {
-  } else if (dailyTask.frequency === 'weekly') {
-    let schedule = moment(startDate).recur();
+    let i = 1;
+    while (schedule.length < 6) {
```
Can you change this while loops to for loops to be more concise.

joe-salomon2 commented:
By "matches" logic, do you mean this previous code:
`    return schedule.matches(startOfDayWithCDSTime) && matchEveryX;
`

TheHollidayInn commented:
Exactly.

joe-salomon2 commented:
So the matches() function is actually part of the moment-recur plugin. Since I am generating the "schedule" manually as an array, rather than a moment-recur object, I won't be able to use the previous matching check.

TheHollidayInn commented:
Alright, let me think it through a bit. We have forked the moment-reccurr library and I have plans on enhancing it to meeting more needs. I'd rather keep logic abstracted on that path.

Can you link your performance measurements/benchmarks to show this change is better?

joe-salomon2 commented:
Oh wow, I was preparing the timing tests for you guys, and it actually wasn't the .recur().every() calculation function that was the culprit, it was the matches() function. I guess that makes sense, since it has to do a match check on all the generated dates.

Let me know how you want to proceed.

TheHollidayInn commented:
The issue should be the next, prev and all functions. How are you measuring? I'll be posting some local tests as well.

joe-salomon2 commented:
I was grabbing timestamps at certain points in the code like below and monitoring the values in Chrome. Local unit tests didn't seem to produce any time issues though.

```
} else if (dailyTask.frequency === 'yearly') {
    let t0 = performance.now();
    let schedule = moment(startDate).recur();

    schedule = schedule.every(dailyTask.everyX).years();
    
    if (options.nextDue) {
      let dates = schedule.fromDate(startOfDayWithCDSTime).next(6);
      let filterDates = dates.filter((momentDate) => {
        let monthDiff = momentDate.years() - moment(startDate).years();
        let matchX = monthDiff % dailyTask.everyX === 0;
        return matchX;
      });
      let t1 = performance.now();
      console.log(t1-t0);
      return filterDates;
    }

    return schedule.matches(startOfDayWithCDSTime);
  }
```
and
```
function generateNextDue (task, user) {
      var options = angular.copy(user);
      options.nextDue = true;
      var t0 = performance.now();
      var nextDueDates = Shared.shouldDo(new Date, task, options);
      var t1 = performance.now();
      console.log(t1 - t0);
```

TheHollidayInn commented:
All the below are averages of 10 tests. 

Matching (doesn't seem to be the problem):
 - everyX 300, 3.0ms
 - everyX 30, 2.7ms
 - everyX 3, 2.6ms

Generating Dates:
 - everyX 300, 25375.0ms
 - everyX 30, 2325.7ms
 - everyX 3, 253.6ms

Local unit tests to show the issue. I am sure Chrome times the tab out after a delay and tries to close.
So, the only fix we needs is generating the next dues. It also seems this issue only happens on yearlies.

TheHollidayInn commented:
Using this code: 
```
if (options.nextDue) {
      let filterDates = [];
      for (let i = 1; i < 6; i += 1) {
        let calcDate = moment(startDate).add(dailyTask.everyX * i, 'years');
        if (calcDate >= startOfDayWithCDSTime) filterDates.push(calcDate);
      }
      return filterDates;
    }
```

Reduces the time to 1ms. So, I think that is all we need for this fix. We can take this idea and apply it to the moment recurr repo to improve performance as well.

joe-salomon2 commented:
I get what you're saying now. Will fix

joe-salomon2 commented:
Writing additional tests now to make sure the calculated values are valid.

missingno5678 committed:
Merge branch 'develop' into dailies-performance-fix

missingno5678 committed:
change only nextDue calculations
add tests to make sure proper nextDue values are calculated
revert schedule.matches logic to original
revert shouldDo.test.js to original

TheHollidayInn commented on test/api/v3/unit/libs/cron.test.js:
```diff
@@ -392,6 +392,118 @@ describe('cron', () => {
       expect(tasksByType.dailys[0].nextDue.length).to.eql(6);
     });
 
+    it('should compute daily nextDue values', () => {
```
These tests should go in the common tests directory. Thanks for writing them.

missingno5678 committed:
Merge branch 'develop' into dailies-performance-fix

missingno5678 committed:
fix monthly nextDue logic
move tests to shouldDo.test.js

missingno5678 committed:
typos

paglias commented on website/common/script/cron.js:
```diff
@@ -207,6 +219,6 @@ export function shouldDo (day, dailyTask, options = {}) {
     return schedule.matches(startOfDayWithCDSTime);
   }
-
+  console.log(2);
```
this should be removed

joe-salomon2 replied:
The latest checkin has it removed. Thanks for reviewing

TheHollidayInn commented on website/common/script/cron.js:
```diff
@@ -110,11 +110,9 @@ export function shouldDo (day, dailyTask, options = {}) {
     for (let [repeatDay, active] of Object.entries(dailyTask.repeat)) {
-      if (!isFinite(DAY_MAPPING_STRING_TO_NUMBER[repeatDay])) continue; // eslint-disable-line no-continue
-      if (active) daysOfTheWeek.push(parseInt(DAY_MAPPING_STRING_TO_NUMBER[repeatDay], 10));
+      if (active && !isNaN(DAY_MAPPING_STRING_TO_NUMBER[repeatDay])) daysOfTheWeek.push(parseInt(DAY_MAPPING_STRING_TO_NUMBER[repeatDay], 10));
```
Why did you change this line?

joe-salomon2 replied:
For some reason, the Object.entries() function was mapping additional functions/properties for dailyTask.repeat, which pushed unnecessary values to daysOfTheWeek. I was using the daysOfTheWeek.length property later on.

I don't have the output handy, but I can generate it again for you if you'd like in a few.

TheHollidayInn replied:
Yea, the isFinite function should fix that though

joe-salomon2 replied:
you're right. I can quickly fix

TheHollidayInn commented:
Looking good. I just had one comment. I'll test locally later and check the performance as well.

missingno5678 committed:
revert to original logic. change not needed

TheHollidayInn commented:
The CI build broke. I am rerunning. Any ideas why the coverage went down?

joe-salomon2 commented:
Weird, let me check on my end.

joe-salomon commented:
Here's the output just from a fresh develop branch. Do you notice anything odd?
https://gist.github.com/joe-salomon/3aa92a874eb8a3f8ae0f45fe6786edd4

joe-salomon commented:
When comparing my two coveralls below, the one for this PR doesn't have the job for 'TEST="test:api-v3" REQUIRES_SERVER=true COVERAGE=true'. Maybe that is why the coverage dropped?
This PR #8767 : https://coveralls.io/builds/11855993
PR #8749: https://coveralls.io/builds/11855636

missingno5678 committed:
Merge branch 'develop' into dailies-performance-fix

missingno5678 committed:
add failure cases

joe-salomon2 commented:
Fixed the issues :)

TheHollidayInn commented:
Great work. This is super good. Eventually, I'd like to move this too our fork of moment recur too. :D

TheHollidayInn commented:
Noted towards your Tier!

Alys commented:
@joe-salomon You now have tier 2 and are part way to tier 3.
